File: golang/warmup/reverse_vowels.go
package warmup

import "unicode"

func reverseVowels(s string) string {
	result := []rune(s)
	start := 0
	end := len(result) - 1

	for start < end {
		for start < end && !isVowel(result[start]) {
			start++
		}

		for start < end && !isVowel(result[end]) {
			end--
		}

		// switch vowels
		result[start], result[end] = result[end], result[start]

		start++
		end--
	}

	return string(result)
}

func isVowel(c rune) bool {
	vowels := map[rune]bool{
		'a': true,
		'e': true,
		'i': true,
		'o': true,
		'u': true,
	}

	v := unicode.ToLower(c)
	if _, ok := vowels[v]; ok {
		return true
	}

	return false
}

File: rust/src/lib/warmup/reverse_vowels.rs
use std::collections::HashSet;

fn is_vowel(c: char) -> bool {
    let vowels: HashSet<char> = "aeiou".chars().collect();
    vowels.contains(&c.to_ascii_lowercase())
}

fn reverse_vowels(s: &str) -> String {
    let mut start = 0;
    let mut end = s.len().saturating_sub(1);
    let mut result: Vec<char> = s.chars().collect();

    while start < end {
        while start < end && !is_vowel(result[start]) {
            start += 1;
        }

        while start < end && !is_vowel(result[end]) {
            end -= 1;
        }

        // switch vowels
        result.swap(start, end);

        start += 1;
        end -= 1;
    }

    result.into_iter().collect()
}

#[cfg(test)]
pub mod test {
    use super::*;

    #[test]
    fn test_reverse_vowels() {
        let test_cases = [
            ("hello", "holle"),
            ("DesignGUrus", "DusUgnGires"),
            ("AEIOU", "UOIEA"),
            ("aA", "Aa"),
            ("", ""),
        ];

        for (s, expected) in test_cases {
            assert_eq!(reverse_vowels(s), expected);
        }
    }
}
